src/cpu/flexcpu/simple_dataflow_thread.cc

```diff
@@ -132,76 +161,153 @@ SDCPUThread::advanceInst(TheISA::PCState next_pc)
-        make_shared<InflightInst>(getThreadContext(), isa, &memIface, seq_num,
-                                  next_pc);
+        make_shared<InflightInst>(getThreadContext(), isa, &memIface,
+                                  seq_num, next_pc, static_inst);
 
     inflightInsts.push_back(dynamic_inst_ptr);
 
-    fetchOffset = 0;
-    attemptFetch(weak_ptr<InflightInst>(dynamic_inst_ptr));
+    DPRINTF(SDCPUInstEvent, "Issuing instruction (seq %d). Buffer now %d\n",
+                            dynamic_inst_ptr->seqNum(), inflightInsts.size());
+
+#if TRACING_ON
+    // Calls new, must delete eventually.
+    dynamic_inst_ptr->traceData(
+        _cpuPtr->getTracer()->getInstRecord(curTick(), this,
+                dynamic_inst_ptr->staticInst(), dynamic_inst_ptr->pcState(),
+                curMacroOp));
+#else
+    dynamic_inst_ptr->traceData(nullptr);
+#endif
+
+    populateDependencies(dynamic_inst_ptr);
+    populateUses(dynamic_inst_ptr);
+
+    weak_ptr<InflightInst> inst = dynamic_inst_ptr;
+    if (dynamic_inst_ptr->isReady()) { // If no dependencies, execute now
+        executeInstruction(inst);
+    } else { // Else, add an event callback to execute when ready
+        dynamic_inst_ptr->addReadyCallback([this, inst](){
+            executeInstruction(inst);
+        });
+    }
+
+    // Try to start the next decode/issue
+    if (hasNextPC()) {
+        // If we are still running and know where to fetch, we should
+        // immediately send the necessary requests to the CPU to fetch the next
+        // instruction
+        advanceInst(getNextPC());
+    } else if (_cpuPtr->hasBranchPredictor()) {
+        if (!remainingBranchPredDepth) {
+            DPRINTF(SDCPUBranchPred, "This control would exceed the branch "
+                                      "prediction depth limit, not requesting "
+                                      "a prediction immediately.\n");
+
+            unpredictedBranches.push_back(dynamic_inst_ptr);
+
+            return;
+        }
+
+        --remainingBranchPredDepth;
+
+        predictCtrlInst(dynamic_inst_ptr);
+    } else {
+        DPRINTF(SDCPUThreadEvent, "Delaying fetch until control instruction's "
+                                  "execution.\n");
+    }
 }
 
 void
-SDCPUThread::attemptFetch(weak_ptr<InflightInst> inst)
+SDCPUThread::attemptFetch(TheISA::PCState next_pc)
 {
-    const shared_ptr<InflightInst> inst_ptr = inst.lock();
-    if (!inst_ptr || inst_ptr->isSquashed()) {
-        // No need to do anything for an instruction that has been squashed.
-        return;
-    }
-
     DPRINTF(SDCPUThreadEvent, "attemptFetch()\n");
 
+    // Save the PC we are trying to fetch from. It's possible that a branch
+    // misprediction will cause us to fetch from a different PC before this
+    // memory request returns. If, when this request is done, the fetch PC
+    // is different, then we need to just ignore this data.
+    fetchPC = next_pc;
```

Comment:
I wonder if that is a safe assumption to make -- that if a request response matches the PC we want, then it's always safe to accept it?